pkg/ip/cidr.go

```diff
@@ -0,0 +1,154 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ip
+
+import (
+	"encoding/binary"
+	"math/big"
+	"net"
+)
+
+// NextIP returns IP incremented by 1, if IP is invalid, return nil
+func NextIP(ip net.IP) net.IP {
+	normalizedIP := normalizeIP(ip)
+	if normalizedIP == nil {
+		return nil
+	}
+
+	i := ipToInt(normalizedIP)
+	return intToIP(i.Add(i, big.NewInt(1)), len(normalizedIP) == net.IPv6len)
+}
+
+// NextIPWithOffset returns IP incremented by offset, if IP is invalid, return nil
+func NextIPWithOffset(ip net.IP, offset int64) net.IP {
+	normalizedIP := normalizeIP(ip)
+	if normalizedIP == nil {
+		return nil
+	}
+
+	i := ipToInt(normalizedIP)
+	return intToIP(i.Add(i, big.NewInt(offset)), len(normalizedIP) == net.IPv6len)
+}
+
+// PrevIP returns IP decremented by 1, if IP is invalid, return nil
+func PrevIP(ip net.IP) net.IP {
+	normalizedIP := normalizeIP(ip)
+	if normalizedIP == nil {
+		return nil
+	}
+
+	i := ipToInt(normalizedIP)
+	return intToIP(i.Sub(i, big.NewInt(1)), len(normalizedIP) == net.IPv6len)
+}
+
+// Cmp compares two IPs, returning the usual ordering:
+// a < b : -1
+// a == b : 0
+// a > b : 1
+// incomparable : -2
+func Cmp(a, b net.IP) int {
+	normalizedA := normalizeIP(a)
+	normalizedB := normalizeIP(b)
+
+	if len(normalizedA) == len(normalizedB) && len(normalizedA) != 0 {
+		return ipToInt(normalizedA).Cmp(ipToInt(normalizedB))
+	}
+
+	return -2
+}
+
+// Distance returns amount of IPs between a and b
+// returns -1 if result is negative
+// returns -2 if result is too large
+func Distance(a, b net.IP) int64 {
+	normalizedA := normalizeIP(a)
+	normalizedB := normalizeIP(b)
+	count := big.NewInt(0).Sub(ipToInt(normalizedB), ipToInt(normalizedA))
+	if !count.IsInt64() {
+		return -2
+	}
+	if count.Cmp(big.NewInt(0)) < 0 {
+		return -1
+	}
+	return count.Int64()
+}
+
+func ipToInt(ip net.IP) *big.Int {
+	return big.NewInt(0).SetBytes(ip)
+}
+
+func intToIP(i *big.Int, isIPv6 bool) net.IP {
+	intBytes := i.Bytes()
+
+	if len(intBytes) == net.IPv4len || len(intBytes) == net.IPv6len {
+		return intBytes
+	}
+
+	if isIPv6 {
+		return append(make([]byte, net.IPv6len-len(intBytes)), intBytes...)
+	}
+
+	return append(make([]byte, net.IPv4len-len(intBytes)), intBytes...)
+}
+
+// normalizeIP will normalize IP by family,
+// IPv4 : 4-byte form
+// IPv6 : 16-byte form
+// others : nil
+func normalizeIP(ip net.IP) net.IP {
+	if ipTo4 := ip.To4(); ipTo4 != nil {
+		return ipTo4
+	}
+	return ip.To16()
+}
+
+// Network masks off the host portion of the IP, if IPNet is invalid,
+// return nil
+func Network(ipn *net.IPNet) *net.IPNet {
+	if ipn == nil {
+		return nil
+	}
+
+	maskedIP := ipn.IP.Mask(ipn.Mask)
+	if maskedIP == nil {
+		return nil
+	}
+
+	return &net.IPNet{
+		IP:   maskedIP,
+		Mask: ipn.Mask,
+	}
+}
+
+// IsBroadcast returns true if provided IP is IPv4 Broadcast ip of the network
+func IsBroadcast(ip net.IP, network *net.IPNet) bool {
+	if network == nil {
```

Comment:
nit: all these checks should end in error.
or function should make assumption that args are correct

technically you can implement `func  GetBroadcastIP(network *net.IPNet) net.IP` which will return the broadcast IP (last ip) of the subnet which is essentially L150-152 of this func

then compare RangeEnd IP to it in allocator.

Reply:
ack, just need to keep in mind that using this and getting false means:

1. its not a boradcast address
2. some error happend

that is false, doesnt tell us much unless we are sure inputs are correct